File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/Tests/TestsUseBehaviorSpec.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class TestsUseBehaviorSpec: QuickSpec {
    override class func spec() {
        Given("A test class") {
            let tests = WoofHarmonize.testCode.classes().withSuffix("Test")

            Then("It is a QuickSpec that reads as Given / When / Then") {
                tests.assertTrue(rule: rule) { test in
                    test.inheritanceTypesNames.contains("QuickSpec")
                        && test.description.contains("Given(")
                        && test.description.contains("Then(")
                }
            }
        }

        Given("A test file") {
            let files = WoofHarmonize.testCode.sources()

            Then("It does not use XCTestCase, sleep, or the raw Quick DSL") {
                files.assertTrue(rule: rule) { file in
                    !file.source.contains("XCTestCase")
                        && !file.source.containsMatch(of: sleepPattern)
                        && !file.source.containsMatch(of: rawDSLPattern)
                }
            }
        }
    }

    private static let sleepPattern = try! NSRegularExpression(pattern: #"\b(sleep|usleep|Thread\.sleep)\("#)
    private static let rawDSLPattern = try! NSRegularExpression(pattern: #"^\s*(describe|context|it)\("#, options: [.anchorsMatchLines])

    private static let rule = Rule(
        description: "Tests are QuickSpecs written as Given / When / Then with beforeEach setup and justBeforeEach actions.",
        rationale: """
            Tests target the ViewModel layer only, which exercises use cases and repositories underneath.
            The BDD structure reads as a spec, and a fresh Container per beforeEach gives every Then a clean slate.
            """,
        fixHint: "Subclass QuickSpec, use Given/When/Then/And blocks, set up in beforeEach, act in justBeforeEach, and advance time with TimeAdvancingFactory(container).tick().",
        badExample: "final class GridViewModelTest: XCTestCase { func testLoads() { sleep(1) } }",
        goodExample: "final class GridViewModelTest: QuickSpec { override class func spec() { Given(\"I open the grid\") { Then(\"...\") { ... } } } }"
    )
}

File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/tests/TestsUseBehaviorSpec.kt
package com.perrystreet.woof.konsist.tests

import com.lemonappdev.konsist.api.ext.list.withNameEndingWith
import com.perrystreet.woof.konsist.Assertions.assertTrue
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class TestsUseBehaviorSpec : BehaviorSpec() {
    init {
        Given("A test class") {
            val tests = KonsistUtils.testCode.classes().withNameEndingWith("Test")

            Then("It extends ViewModelBehaviorSpec and reads as Given / When / Then") {
                tests.assertTrue(message = Message) { test ->
                    test.hasParentWithName("ViewModelBehaviorSpec") &&
                        test.text.contains("Given(") &&
                        test.text.contains("Then(")
                }
            }
        }

        Given("A test file") {
            val files = KonsistUtils.testCode.files

            Then("It does not use beforeTest, Thread.sleep, or other spec styles") {
                files.assertTrue(message = Message) { file ->
                    !file.text.contains("beforeTest") &&
                        !file.text.contains("Thread.sleep") &&
                        !file.hasImport { it.name.contains("DescribeSpec") || it.name.contains("FunSpec") || it.name.contains("StringSpec") }
                }
            }
        }
    }

    private companion object {
        private val Message = LintRuleMessage(
            rule = "Tests are ViewModelBehaviorSpecs written as Given / When / Then with beforeEach setup.",
            why = """
                Tests target the ViewModel layer only, which exercises use cases and repositories underneath.
                The BDD structure reads as a spec, and InstancePerLeaf plus beforeEach gives every Then a clean slate.
            """.trimIndent(),
            howToFix = "Extend ViewModelBehaviorSpec, use Given/When/Then/And blocks, set up in beforeEach, and advance time with TimeAdvancingFactory().tick().",
            badExample = "class GridViewModelTest : FunSpec({ test(\"loads\") { Thread.sleep(100) } })",
            goodExample = "class GridViewModelTest : ViewModelBehaviorSpec() { init { Given(\"I open the grid\") { Then(\"...\") { ... } } } }",
        )
    }
}
